lib/http/features/auto_deflate.rb

```diff
@@ -36,6 +49,11 @@ def size
           @compressed.size
         end
 
+        def read(*a)
+          compress_all! unless @compressed
+          @compressed.read(*a)
+        end
+
```

Comment:
I was trying to get the `CompressedBody` to quack like a `Request::Body` so that this branch wouldn't fail: https://github.com/httprb/http/blob/master/lib/http/request/body.rb#L20 

There's probably a better approach, but I wasn't really sure how this code in `Request::Body` was for, so I wanted my change to remain low-impact.

Reply:
I see. But isn't `CompressedBody` wrapping `Request::Body`, not the other way around?

If I'm wrong, then in `Request::Body` we can modify the branch where it checks if body is `Enumerable` to just checking if it responds to `#each` instead (we don't actually need the object to be Enumerable, we're just using `#each` anyway).

Reply:
That was the first thing I tried, actually, but several other tests failed, and when I started digging in I decided that would be a larger change. Also, since `Request::Body#size` gets used to set the `Content-Length` header, I was worried that if it was passed an array-like enumerable like `["foo", "bar"]`, then `size` would be 2 instead of 6. I don't know if its ever possible for `Body` to ever have an array source, but it seemed easier to make `AutoDeflate` quack like a normal body instead.

Reply:
The arrangement in this PR makes sense to me except for `Request::Body` being the actual compressed body (i.e. on the wire) and `CompressedBody` actually being the decompressed body, though that feels like a silly bikeshed/nit

Reply:
Ok, I see now what's the problem. The `Request::Body` is being wrapped by `DeflatedBody` and passed as `:body` to `Request.new`, which then wraps `Request::Body` around it again. This causes an issue when trying to determine size, because the outer `Request::Body` isn't able to determine the size, even though it's `CompressedBody` source responds to `#size`. We should somehow try to avoid it, that `Request#body` ends up being `CompressedBody`.

I'm reluctant to have this PR merged as is because it loses on-the-fly compression, as `CompressedBody#each` is never called since `Request::Body` will  pick the new `CompressedBody#read`.

Reply:
@janko-m Ok, I tweaked it a bit so that `CompressedBody` inherits from `Request::Body`, and the `Request` doesn't try to wrap the body in a body if its already a body.